src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,632 @@
+         *      is why it is skipped for now. */
+        fprintf(f,
+                "After=graphical-session.target\n");
+
+        fprintf(f,
+                "\n[Service]\n"
+                "Type=simple\n"
+                "ExecStart=:%s\n"
+                "Restart=no\n"
+                "TimeoutSec=5sec\n"
+                "Slice=apps.slice\n",
+                exec_start);
+
+        /* Generate an ExecCondition to check $XDG_CURRENT_DESKTOP */
+        if (!strv_isempty(service->only_show_in) || !strv_isempty(service->not_show_in)) {
+                _cleanup_free_ char *only_show_in = NULL;
+                _cleanup_free_ char *not_show_in = NULL;
+                _cleanup_free_ char *e_only_show_in = NULL;
+                _cleanup_free_ char *e_not_show_in = NULL;
+
+                only_show_in = strv_join(service->only_show_in, ":");
+                not_show_in = strv_join(service->not_show_in, ":");
+                if (!only_show_in || !not_show_in)
+                        return log_oom();
+
+                e_only_show_in = cescape(only_show_in);
+                e_not_show_in = cescape(not_show_in);
+                if (!e_only_show_in || !e_not_show_in)
+                        return log_oom();
+
+                /* Just assume the values are reasonably sane */
+                fprintf(f,
+                        "ExecCondition=" ROOTLIBEXECDIR "/systemd-xdg-autostart-condition \"%s\" \"%s\"\n",
+                        e_only_show_in,
+                        e_not_show_in);
+        }
+
+        /* Generate an ExecCondition for GNOME autostart condition */
+        if (!isempty(service->kde_autostart_condition)) {
+                _cleanup_free_ char *gnome_autostart_condition_path = NULL;
+                _cleanup_free_ char *e_autostart_condition = NULL;
+
+                if (find_binary("gnome-systemd-autostart-condition", &gnome_autostart_condition_path) < 0) {
+                        e_autostart_condition = cescape(service->autostart_condition);
+                        if (!e_autostart_condition)
+                                return log_oom();
+
+                         fprintf(f,
+                                 "ExecCondition=%s --condition \"%s\"\n",
+                                 gnome_autostart_condition_path,
+                                 e_autostart_condition);
+                 } else {
+                         log_info_errno(r, "gnome-systemd-autostart-condition not found: %m");
+                         fprintf(f, "# ExecCondition for GNOME skipped due to missing binary\n");
+                }
+        }
+
+        /* Generate an ExecCondition for KDE autostart condition */
+        if (!isempty(service->kde_autostart_condition)) {
+                _cleanup_free_ char *kde_autostart_condition_path = NULL;
+                _cleanup_free_ char *e_kde_autostart_condition = NULL;
+
+                if (find_binary("kde-systemd-start-condition", &kde_autostart_condition_path) < 0) {
```

Comment:
why use find_binary() here? just hardcode the paths here, via macro from build system

Reply:
@unusual-thoughts, do you have an opinion on this? I am completely fine with hardcoding the path, but not sure where I can expect the binary to live.

Reply:
we know where our own binaries live, we install them with our build scripts after all...

Reply:
oh, i see now, it's not actually our binary. ok, leave it in then

Reply:
We know where systemd's binaries live, but this kde-systemd-start-condition is shipped on the KDE side which systemd won't know about.

The idea was instead of supporting large amounts of KDE code in this patch, we delegate custom desktop-specific conditionals to an area the desktop controls.

We definitely have devs who install into random prefixes.

Reply:
Yeah, it is the integration point for desktops. Not great, but I don't see a better way of doing this.

And, to scare you … I actually have working code that creates files/directories for the conditions in `XDG_RUNTIME_DIR` and a small daemon that starts/stops services when their condition changes. Don't really like it, it does work though. Not sure we'll ever use it.